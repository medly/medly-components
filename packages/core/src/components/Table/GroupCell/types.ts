import { GridTemplateProps } from '../types';
export interface GroupCellProps extends GridTemplateProps {
    hidden: boolean;
    isTitleCell?: boolean;
}
