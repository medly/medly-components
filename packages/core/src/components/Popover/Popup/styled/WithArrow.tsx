import { css } from 'styled-components';
import { PopoverPopupProps } from '../types';

const getPositionWithArrow = ({ placement }: PopoverPopupProps) => {
    switch (placement) {
        case 'top':
            return css`
                box-shadow: 0 0.4rem 1.6rem rgba(176, 188, 200, 0.5);
                &::after {
                    transform: translateX(-50%);
                    top: 100%;
                    left: 50%;
                }
            `;
        case 'top-start':
            return css`
                box-shadow: 0 0.4rem 1.6rem rgba(176, 188, 200, 0.5);
                &::after {
                    top: 100%;
                    left: 2rem;
                }
            `;
        case 'top-end':
            return css`
                box-shadow: 0 0.4rem 1.6rem rgba(176, 188, 200, 0.5);
                &::after {
                    top: 100%;
                    right: 2rem;
                }
            `;
        case 'right':
            return css`
                box-shadow: -0.4rem 0 1.6rem rgba(176, 188, 200, 0.5);
                &::after {
                    transform: translateY(-50%) rotate(90deg);
                    top: 50%;
                    right: calc(100% - 0.3rem);
                }
            `;
        case 'right-start':
            return css`
                box-shadow: -0.4rem 0 1.6rem rgba(176, 188, 200, 0.5);
                &::after {
                    transform: rotate(90deg);
                    top: 2rem;
                    right: calc(100% - 0.3rem);
                }
            `;
        case 'right-end':
            return css`
                box-shadow: -0.4rem 0 1.6rem rgba(176, 188, 200, 0.5);
                &::after {
                    transform: rotate(90deg);
                    bottom: 2rem;
                    right: calc(100% - 0.3rem);
                }
            `;
        case 'bottom':
            return css`
                box-shadow: 0 -0.4rem 1.6rem rgba(176, 188, 200, 0.5);
                &::after {
                    transform: translateX(-50%) rotate(180deg);
                    left: 50%;
                    bottom: 100%;
                }
            `;
        case 'bottom-start':
            return css`
                box-shadow: 0 -0.4rem 1.6rem rgba(176, 188, 200, 0.5);
                &::after {
                    transform: rotate(180deg);
                    left: 2rem;
                    bottom: 100%;
                }
            `;
        case 'bottom-end':
            return css`
                box-shadow: 0 -0.4rem 1.6rem rgba(176, 188, 200, 0.5);
                &::after {
                    transform: rotate(180deg);
                    right: 2rem;
                    bottom: 100%;
                }
            `;
        case 'left':
            return css`
                box-shadow: 0.4rem 0 1.6rem rgba(176, 188, 200, 0.5);
                &::after {
                    transform: translateY(-50%) rotate(270deg);
                    top: 50%;
                    left: calc(100% - 0.3rem);
                }
            `;
        case 'left-start':
            return css`
                box-shadow: 0.4rem 0 1.6rem rgba(176, 188, 200, 0.5);
                &::after {
                    transform: rotate(270deg);
                    top: 2rem;
                    left: calc(100% - 0.3rem);
                }
            `;
        case 'left-end':
            return css`
                box-shadow: 0.4rem 0 1.6rem rgba(176, 188, 200, 0.5);
                &::after {
                    transform: rotate(270deg);
                    bottom: 2rem;
                    left: calc(100% - 0.3rem);
                }
            `;
    }
};

export const getStyleWithArrow = css<PopoverPopupProps>`
    border-radius: 0.8rem;
    padding: 1.5rem 2rem;

    &::after {
        content: '';
        width: 0;
        height: 0;
        z-index: 4;
        position: absolute;
        border-left: 0.8rem solid transparent;
        border-right: 0.8rem solid transparent;
        border-top: 1rem solid ${({ bgColor }) => bgColor};
    }

    ${getPositionWithArrow}
`;
