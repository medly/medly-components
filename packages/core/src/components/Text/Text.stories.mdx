import { Text } from './Text';
import { Preview, Story, Meta, Props } from '@storybook/addon-docs/blocks';
import { Basic, VerticalAlign } from './Text.stories';

<Meta title="Core" component={Text} />

# Text

The `Text` component displays text efficiently. It supports nesting, styling, and touch handling. 

<Preview>
    <Story name="Text">{Basic()}</Story>
</Preview>

## Text with different Variants

You can use the font variant to decide the font style (`font-size`,`font-weight`,`line-height`,`letter-spacing`) as well as the `html tag`. If you use the `textWeight` prop as `Strong`|`ExtraStrong`, then both varaints `body1` and `body2` will use the `strong` html tag.

| Variant | Html Tag |
| ------- | -------- |
| h1      | h1       |
| h2      | h2       |
| h3      | h3       |
| h4      | h4       |
| h5      | h5       |
| body1   | span     |
| body2   | span     |

<Preview>
    <VerticalAlign>
        <Text textVariant="h1">Header 1 / Open Sans - Medium</Text>
        <Text textVariant="h2">Header 2 / Open Sans - Medium</Text>
        <Text textVariant="h3">Header 3 / Open Sans - Regular</Text>
        <Text textVariant="h4">Header 4 / Open Sans - Strong</Text>
        <Text textVariant="h5">HEADER 5 / Open Sans - Strong</Text>
        <Text textVariant="body1">Body 1 / Open Sans - Regular</Text>
        <Text textVariant="body2">Body 2 / Open Sans - Regular</Text>
    </VerticalAlign>
</Preview>

## Text with different fontWeight

You can use the different font weights to define the thinness or thickness of a font like mentioned below.

<Preview>
    <VerticalAlign>
        <Text textWeight="Light">Body 2 / Open Sans - Light</Text>
        <Text textWeight="Regular">Body 2 / Open Sans - Regular</Text>
        <Text textWeight="Medium">Body 2 / Open Sans - Medium</Text>
        <Text textWeight="Strong">Body 2 / Open Sans - Strong</Text>
        <Text textWeight="ExtraStrong">Body 2 / Open Sans - Extra Strong</Text>
    </VerticalAlign>
</Preview>

## Props

This component offers the following props mentioned in the table below:

<Props of={Text} />
