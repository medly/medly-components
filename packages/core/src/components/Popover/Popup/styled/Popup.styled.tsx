import { fullHeight, fullWidth } from '@medly-components/utils';
import styled, { css } from 'styled-components';
import { Placement } from '../../types';
import { StyledPopoverPopupProps } from '../types';
import { getStyleWithArrow } from './WithArrow';

export const getPosition = ({ placement, distance = '0px' }: { placement: Placement; distance?: string }) => {
    switch (placement) {
        case 'top-start':
            return css`
                bottom: calc(100% + ${distance});
                left: 0%;
            `;
        case 'top':
            return css`
                bottom: calc(100% + ${distance});
                left: 50%;
                transform: translateX(-50%);
            `;
        case 'top-end':
            return css`
                bottom: calc(100% + ${distance});
                right: 0%;
            `;
        case 'right-start':
            return css`
                top: 0%;
                left: calc(100% + ${distance});
            `;
        case 'right':
            return css`
                left: calc(100% + ${distance});
                top: 50%;
                transform: translateY(-50%);
            `;
        case 'right-end':
            return css`
                bottom: 0;
                left: calc(100% + ${distance});
            `;
        case 'bottom-start':
            return css`
                top: calc(100% + ${distance});
                left: 0%;
            `;
        case 'bottom':
            return css`
                top: calc(100% + ${distance});
                left: 50%;
                transform: translateX(-50%);
            `;
        case 'bottom-end':
            return css`
                top: calc(100% + ${distance});
                right: 0%;
            `;
        case 'left-start':
            return css`
                right: calc(100% + ${distance});
                top: 0%;
            `;
        case 'left':
            return css`
                right: calc(100% + ${distance});
                top: 50%;
                transform: translateY(-50%);
            `;
        case 'left-end':
            return css`
                right: calc(100% + ${distance});
                bottom: 0%;
            `;
    }
};

export const PopupStyled = styled('div')<StyledPopoverPopupProps>`
    z-index: 4;
    position: absolute;
    background-color: ${({ bgColor }) => bgColor};

    ${props => props.fullWidth && fullWidth()};
    ${props => props.fullHeight && fullHeight()};

    ${props => getPosition({ placement: props.placement, distance: props.withArrow ? '1.5rem' : props.distance })}
    ${({ withArrow }) => withArrow && getStyleWithArrow}
`;
