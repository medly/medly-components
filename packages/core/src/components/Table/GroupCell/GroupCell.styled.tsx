import { styled } from '@medly-components/utils';
import Text from '../../Text';
import { getBorder } from '../Table.styled';
import { Props } from './types';

export const GroupCell = styled('div').attrs(({ gridTemplateColumns }: Props) => ({
    style: {
        gridTemplateColumns
    }
}))<Props>`
    display: grid;
    position: relative;
    opacity: ${({ hidden }) => (hidden ? '0' : '1')};
    width: ${({ hidden }) => (hidden ? '0' : '100%')};
    height: ${({ hidden }) => (hidden ? '0' : '100%')};
    &::after {
        ${({ isTitleCell }) => isTitleCell && getBorder('right')}
    }
    padding: 0;
`;

export const GroupCellTitle = styled(Text.Style)`
    position: relative;
    text-align: center;
    padding: 0.9rem 1.6rem 1rem;
    grid-column: -1/1;
    text-overflow: initial;
    overflow: auto;
    white-space: normal;
    color: ${({ theme }) => theme.table.header.cell.textColor.default};
    &::after {
        ${getBorder('bottom')}
    }
`;
