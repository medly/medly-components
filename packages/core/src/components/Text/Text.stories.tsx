import { FontVariants, FontWeights } from '@medly-components/theme';
import { boolean, color, select } from '@storybook/addon-knobs';
import styled from 'styled-components';
import { Text } from './Text';

const variantOptions: FontVariants[] = ['h1', 'h2', 'h3', 'h4', 'h5', 'body1', 'body2'];
const weightOptions: FontWeights[] = ['Light', 'Regular', 'Medium', 'Strong'];

export const Basic = () => (
    <Text
        fullWidth={boolean('Full Width', false)}
        uppercase={boolean('Upper Case', false)}
        lineThrough={boolean('Line through', false)}
        textColor={color('Color', '#012040')}
        textVariant={select('Text Variant', variantOptions, 'body2')}
        textWeight={select('Text Weight', weightOptions, 'Regular')}
    >
        Lorem ipsum dolor sit, amet consectetur adipisicing elit.
    </Text>
);

export const VerticalAlign = styled.div`
    width: 100%;
    display: flex;
    flex-direction: column;
    & > * {
        margin-bottom: 8px;
    }
`;
