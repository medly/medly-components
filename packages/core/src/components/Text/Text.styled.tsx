import { defaultTheme } from '@medly-components/theme';
import { WithThemeProp } from '@medly-components/utils';
import styled, { css } from 'styled-components';
import { TextProps } from './types';

const uppercase = () => css`
    text-transform: uppercase;
`;

const lineThrough = () => css`
    text-decoration: line-through;
`;
const getTextStyle = ({ theme, ...props }: TextProps & WithThemeProp) => {
    const { textColor, textVariant, textWeight, textAlign } = props,
        { defaults, weights, variants } = theme.font,
        { fontSize, fontWeight, lineHeight, letterSpacing } = variants[textVariant || defaults.variant];

    return css`
        margin: 0;
        color: ${textColor};
        font-size: ${fontSize};
        font-weight: ${weights[textWeight || fontWeight]};
        letter-spacing: ${letterSpacing};
        line-height: ${lineHeight};
        text-align: ${textAlign};
    `;
};

export const TextStyled = styled('span')<TextProps>`
    ${getTextStyle}

    ${({ fullWidth, theme }) =>
        fullWidth &&
        css`
            display: block;
            margin: ${theme.spacing.S2} 0;
        `}

    ${props => props.uppercase && uppercase()};
    ${props => props.lineThrough && lineThrough()};
    display: ${({ as }) => (as === 'strong' || as === 'span' || as === 'a') && 'inline-block'};
`;

TextStyled.defaultProps = {
    theme: defaultTheme
};
