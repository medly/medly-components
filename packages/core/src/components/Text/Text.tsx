import { WithStyle } from '@medly-components/utils';
import React, { FC, useMemo } from 'react';
import { TextStyled } from './Text.styled';
import { TextProps } from './types';

export const Text: FC<TextProps> & WithStyle = React.memo(
    React.forwardRef(({ as, textVariant, textWeight, textAlign, children, ...restProps }, ref) => {
        const componentType = useMemo(() => {
            let type: keyof JSX.IntrinsicElements | React.ComponentType<any> = (textVariant as keyof JSX.IntrinsicElements) || 'span';

            if (
                !textVariant ||
                textVariant === 'body1' ||
                textVariant == 'body2' ||
                textVariant == 'body3' ||
                textVariant == 'button1' ||
                textVariant == 'button2'
            ) {
                type = textWeight === 'Strong' || textWeight === 'ExtraStrong' ? 'strong' : 'span';
            }

            return type;
        }, [textWeight, textVariant]);
        return (
            <TextStyled {...{ textVariant, textWeight, textAlign, ...restProps }} as={as || componentType} ref={ref}>
                {children}
            </TextStyled>
        );
    })
);

Text.defaultProps = {
    lineThrough: false,
    uppercase: false,
    fullWidth: false,
    textColor: 'inherit',
    textAlign: 'initial'
};
Text.displayName = 'Text';
Text.Style = TextStyled;
