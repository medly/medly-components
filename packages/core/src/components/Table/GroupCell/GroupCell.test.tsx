import { render } from '@test-utils';
import Text from '../../Text';
import { GroupCell, GroupCellTitle } from './GroupCell.styled';

describe('GroupCell', () => {
    it('should render properly', () => {
        const { container } = render(
            <GroupCell hidden={false}>
                <Text>Dummy</Text>
            </GroupCell>
        );
        expect(container).toMatchSnapshot();
    });

    it('should not render hidden prop is given', () => {
        const { container } = render(
            <GroupCell hidden>
                <Text>Dummy</Text>
            </GroupCell>
        );
        expect(container).toMatchSnapshot();
    });

    it('should render GroupCellTitle properly', () => {
        const { container } = render(
            <GroupCellTitle>
                <Text>Dummy</Text>
            </GroupCellTitle>
        );
        expect(container).toMatchSnapshot();
    });

    it('should render GroupCell with right border if it is a title cell', () => {
        const { container } = render(
            <GroupCell isTitleCell hidden={false}>
                <Text>Dummy</Text>
            </GroupCell>
        );
        expect(container).toMatchSnapshot();
    });
});
