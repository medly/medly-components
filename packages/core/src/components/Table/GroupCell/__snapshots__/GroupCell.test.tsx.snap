// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GroupCell should not render hidden prop is given 1`] = `
.c1 {
  margin: 0;
  color: inherit;
  font-size: 1.4rem;
  font-weight: 400;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  line-height: 2.2rem;
  text-align: initial;
  display: inline-block;
}

.c0 {
  display: grid;
  position: relative;
  opacity: 0;
  width: 0;
  height: 0;
  padding: 0;
}

<div>
  <div
    class="c0"
    hidden=""
  >
    <span
      class="c1"
    >
      Dummy
    </span>
  </div>
</div>
`;

exports[`GroupCell should render GroupCell with right border if it is a title cell 1`] = `
.c1 {
  margin: 0;
  color: inherit;
  font-size: 1.4rem;
  font-weight: 400;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  line-height: 2.2rem;
  text-align: initial;
  display: inline-block;
}

.c0 {
  display: grid;
  position: relative;
  opacity: 1;
  width: 100%;
  height: 100%;
  padding: 0;
}

.c0::after {
  content: '';
  position: absolute;
  pointer-events: none;
  background-color: #C7D0D8;
  height: calc(100% - 3.2rem);
  width: 1px;
  right: 0;
  top: 50%;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
}

<div>
  <div
    class="c0"
  >
    <span
      class="c1"
    >
      Dummy
    </span>
  </div>
</div>
`;

exports[`GroupCell should render GroupCellTitle properly 1`] = `
.c1 {
  margin: 0;
  color: inherit;
  font-size: 1.4rem;
  font-weight: 400;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  line-height: 2.2rem;
  text-align: initial;
  display: inline-block;
}

.c0 {
  margin: 0;
  font-size: 1.4rem;
  font-weight: 400;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  line-height: 2.2rem;
  position: relative;
  text-align: center;
  padding: 0.9rem 1.6rem 1rem;
  grid-column: -1/1;
  text-overflow: initial;
  overflow: auto;
  white-space: normal;
  color: #435465;
}

.c0::after {
  content: '';
  position: absolute;
  pointer-events: none;
  background-color: #C7D0D8;
  height: 1px;
  width: calc(100% - 3.2rem);
  left: 50%;
  bottom: 0;
  -webkit-transform: translateX(-50%);
  -ms-transform: translateX(-50%);
  transform: translateX(-50%);
}

<div>
  <span
    class="c0"
  >
    <span
      class="c1"
    >
      Dummy
    </span>
  </span>
</div>
`;

exports[`GroupCell should render properly 1`] = `
.c1 {
  margin: 0;
  color: inherit;
  font-size: 1.4rem;
  font-weight: 400;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  line-height: 2.2rem;
  text-align: initial;
  display: inline-block;
}

.c0 {
  display: grid;
  position: relative;
  opacity: 1;
  width: 100%;
  height: 100%;
  padding: 0;
}

<div>
  <div
    class="c0"
  >
    <span
      class="c1"
    >
      Dummy
    </span>
  </div>
</div>
`;
